common/src/verify.rs

```diff
@@ -75,27 +78,1347 @@ pub fn verify_finalization_proof(
+        csv.apply_commit(&Commit::Block(BlockHeader {
+            author: validator_keypair[0].0.clone(),
+            prev_block_finalization_proof: generate_unanimous_finalization_proof(
+                &validator_keypair,
+                &csv.header,
+            ),
+            previous_hash: Commit::Block(csv.header.clone()).to_hash256(),
+            height: csv.header.height + 1,
+            timestamp: 2,
+            commit_hash: csv.commit_hash,
+            tx_merkle_root: OneshotMerkleTree::create(vec![generate_empty_transaction_commit(
+                &validator_keypair,
+                0,
+                0,
+            )
+            .to_hash256()])
+            .root(),
+            chat_merkle_root: Hash256::zero(),
+            repository_merkle_root: Hash256::zero(),
+            validator_set: validator_keypair
+                .iter()
+                .map(|(public_key, _)| (public_key.clone(), 1))
+                .collect(),
+            version: "0.0.0".to_string(),
+        }))
+        .unwrap_err();
+    }
+
+    #[test]
+    /// Test the case where the block commit is invalid because the commit hash is invalid.
+    fn invalid_block_commit_with_invalid_commit_hash() {
+        let (validator_keypair, _, mut csv) = setup_test(3);
+        // Apply agenda commit
+        let agenda_hash_value = calculate_agenda_hash(csv.phase.clone(), csv.header.height);
+        let agenda: Agenda = Agenda {
+            author: validator_keypair[0].0.clone(),
+            timestamp: 1,
+            hash: agenda_hash_value,
+        };
+        csv.apply_commit(&generate_agenda_commit(&agenda)).unwrap();
+        // Apply agenda-proof commit
+        csv.apply_commit(&generate_agenda_proof_commit(
+            &validator_keypair,
+            &agenda,
+            agenda_hash_value,
+        ))
+        .unwrap();
+        // Apply block commit with invalid commit hash
+        csv.apply_commit(&Commit::Block(BlockHeader {
+            author: validator_keypair[0].0.clone(),
+            prev_block_finalization_proof: generate_unanimous_finalization_proof(
+                &validator_keypair,
+                &csv.header,
+            ),
+            previous_hash: Commit::Block(csv.header.clone()).to_hash256(),
+            height: csv.header.height + 1,
+            timestamp: 2,
+            commit_hash: Hash256::zero(),
+            tx_merkle_root: OneshotMerkleTree::create(vec![]).root(),
+            chat_merkle_root: Hash256::zero(),
+            repository_merkle_root: Hash256::zero(),
+            validator_set: validator_keypair
+                .iter()
+                .map(|(public_key, _)| (public_key.clone(), 1))
+                .collect(),
+            version: "0.0.0".to_string(),
+        }))
+        .unwrap_err();
+    }
+
+    #[test]
+    /// Test the case where the block commit is invalid because block commit already exists.
+    fn multiple_block_commits() {
```

Comment:
Isn't it just a phase violation?

Reply:
Yes it is also a kind of phase mismatch.
There are also other similar tests such as `multiple_agendas` and `multiple_agenda_proofs`.
Should I rename them all to `phase_mistmatch_for...`?

Reply:
Yeah rename all please.
Btw I think we don't need thorough tests for every mismatch cases because it's trivially handled by the wildcard match.
Let's keep them for now but I don't think we don't need any more mismatch tests, FYI.

Reply:
fixed.